### DISCUSSION TITLE
CMU 15-418/Stanford CS149: Parallel Computing - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#138] [Announcements]
# CMU 15-418/Stanford CS149: Parallel Computing - CS自学指南

CS自学指南

https://csdiy.wiki/%E5%B9%B6%E8%A1%8C%E4%B8%8E%E5%88%86%E5%B8%83%E5%BC%8F%E7%B3%BB%E7%BB%9F/CS149/

**Comment by AlphaHorologii** [comment]
想问下这是大学高年级的课程还是研究生课 感觉太硬核了

**Reply by ming-xuan-z** [reply]
15418前置课程只有15213，15213scs的学生一般是大一下上。所以一般418是大二或者大三上。

**Reply by dengyunsheng250** [reply]
这个有交流群不？

**Reply by suiyu0823** [reply]
同问

**Comment by ZJFuYuhang** [comment]
B站有视频，有字幕对听力不好的小伙伴会友好一点[ 网址 ](https://www.bilibili.com/video/BV16k4y1z7z9?p=4)

**Comment by polar9527** [comment]
Is Parallel Programming Hard, And, If So, What Can You Do About It? 
https://mirrors.edge.kernel.org/pub/linux/kernel/people/paulmck/perfbook/perfbook.html
https://mirrors.edge.kernel.org/pub/linux/kernel/people/paulmck/perfbook/perfbook.html

**Reply by polar9527** [reply]
https://arxiv.org/abs/1701.00854

**Comment by suiyu0823** [comment]
请问这个lab环境怎么搭建呢。好像没找到

**Reply by ming-xuan-z** [reply]
应该是跟213的环境是一样的，

**Comment by RunningVic** [comment]
在上并行计算之前，最好先学一下哪些课呢？求上过的大佬回答。

**Reply by ming-xuan-z** [reply]
15418上之前最好上15213和15210

**Reply by usher233** [reply]
请问15210的lectures在哪里可以看，在Panopto上没有找到

**Reply by Chenxuan-Yao** [reply]
同求15210

**Comment by Yingfan99327** [comment]
请问有什么好的C++多线程编程的课程推荐吗

**Comment by star-hengxing** [comment]
把今年 cs149 作业改了一下，能跑在 Windows了
https://github.com/star-hengxing/cs149-xmake

**Comment by bingshunjishi** [comment]
这课程真的太硬核了 不建议在15-213看完之后直接冲 推荐路线15-213 -> 15-162 -> 15-445之后再刷15-418

**Reply by ming-xuan-z** [reply]
建议直接冲，462和445上了队418也没什么帮助。相反cmu一般都是先上418再上410或者445的

**Reply by ie925155** [reply]
15-162是筆誤吧, 你想說的是15-462嗎？ 15-462跟這門課有關係嗎？

**Reply by bingshunjishi** [reply]
您好 感谢纠正 应该是CS162 berkeley的操作系统课程 可能因为我本人太菜 导致刷15-418的lab感到特别吃力 事实上418的前置课程确实只需要15-213或者cs61c 因人而异吧 这个路线比较适合我

**Comment by FireSK7** [comment]
“前段时间大热的 Spark 框架等等。真正意义上将理论与实践完美地结合在了一起。”你在逗我

**Reply by TwinklerG** [reply]
远古课程了

**Comment by wangzhengquan** [comment]
环境怎么搭建

**Comment by wuyafei** [comment]
推荐一个翻译效果不错的版本：https://www.bilibili.com/video/BV1Xz4y1p7ZN/?vd_source=2b9f6fcc0b53b5f5e1607a9e3e8c83bd

**Comment by walker-ai** [comment]
想问下15418或者cs149有自动评分系统吗，怎么知道自己做的对不对呢，找了一圈没找到，418的autolab好像cmu本校学生才能用

**Reply by 0AyanamiRei** [reply]
同问一下.

**Comment by CloudJ1603** [comment]
作业的链接好像都挂了呀。。

**Comment by LeonCai1** [comment]
刚创建的qq群，找一个一起做项目的队友～：552410497

**Comment by PrecureKamiyu** [comment]
Slides 的字体有点难看，哈哈

**Comment by chalengr** [comment]
想知道有没有人指导为什么我在wsl上测第一个作业的3线程是递增的不是递减的

**Reply by Just-lose-it** [reply]
可能一开始已经找到了正确的方法hh

**Comment by Dusk-Hermit** [comment]
lab3 cuda的实验，如果自己笔记本有n卡，可以直接用nvidia/cuda的docker镜像，lab3源仓库拖进去就能make
https://hub.docker.com/r/nvidia/cuda/tags?page=&page_size=&ordering=&name=deve

**Comment by exile0627** [comment]
15-418最新能找到的lecture video应该是Fall 2019的

**Reply by xllhxd** [reply]
请问哪里有2019fall的视频呢？我只能找到csapp的作者讲的2018spring的版本

**Reply by JFKTRSDAKIDHIKA** [reply]
YT上有23fall的lecture

**Reply by xllhxd** [reply]
b 站有cmu 2023 的，但不知道是不是真的2023，课件用的2020的，但讲课老师不是2020的老师

**Reply by minwang-ai** [reply]
求2023的视频

**Comment by Lincaxiao** [comment]
15418和149与UIUC的ece408有什么区别吗？

**Comment by xuOAO** [comment]
15418 spring16的网站打不开了，有好兄弟有ppt么

**Reply by xuOAO** [reply]
不用了，感觉23年的ppt和16年的没啥区别

**Reply by lshAlgorithm** [reply]
网站确实打不开了吗？
我还是挺想看看ppt下面的学生讨论的，感觉帮助很大。老师也会留些思考题在上面。有别的渠道能看到comments吗？

**Reply by xuOAO** [reply]
用wayback machine可以，上面有记录以前时刻的网站，可以看到讨论

**Reply by lshAlgorithm** [reply]
感谢！ヽ(* ﾟ д ﾟ)ノ

**Comment by ytoso** [comment]
CMU 15-418 Spring 2017
- 课程视频：https://www.youtube.com/watch?v=f0XhfVziYGM&list=PLi2nduiHpoHgoR8TkoRS2T0c0ENog_yUc
- 课程网站：http://web.archive.org/web/20170616064429/http://15418.courses.cs.cmu.edu/spring2017/home

**Comment by ytoso** [comment]
如果大家用wsl或者虚拟机的话，记得在配置文件里把可用processors数量拉到16，不然加不了速。

比如assignment 1，我的wsl配置processors=2，所以不能同时跑16个threads，应该改成processors=16

**Reply by ada-ys325** [reply]
想问问作业内容是什么？

**Reply by Kriston-12** [reply]
请问一下各位佬们，第一个lab 运行 ./mandelbrot， 报错Wrote image file mandelbrot-thread.ppm，但是我看lab document里面说直接run就可以了。请问大家有遇到过这个问题吗。环境是wsl ubuntu22.04. 请各位佬帮帮孩子
$ ./mandelbrot
[mandelbrot serial]: [311.457] ms
Wrote image file mandelbrot-serial.ppm
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
[mandelbrot thread]: [0.140] ms
Wrote image file mandelbrot-thread.ppm
Mismatch : [0][143], Expected : 1, Actual : 0
Error : Output from threads does not match serial outp

**Reply by Kriston-12** [reply]
已解决

**Comment by Cansal7159** [comment]
有无大佬知道，这个CMU 15-440/640 Distributed Systems怎么样，想入手了

**Comment by ytoso** [comment]
我整理了[CMU15418 Spring 2016 Slides](https://github.com/yvtongsong/CMU-15418-Standford-CS149)，配合课程视频食用更香~

**Reply by ytoso** [reply]
网址失效已更新～

**Comment by JustinJin04** [comment]
笔记本cpu型号为11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz，环境是wsl。从lscpu来看是支持hyperthreading的（Thread(s) per core是2），但是在lab1中指定8个thread跑在4个core上却始终不能达到超过4x的加速比。请问有大佬知道原因吗？

**Reply by ytoso** [reply]
你能看到我之前的[评论](https://github.com/PKUFlyingPig/cs-self-learning/discussions/138#discussioncomment-10574890)吗

**Reply by peijli** [reply]
起码这个学期CS149的课程环境，是AWS上的64位ARM处理器……大部分作业在x86的系统上性能会差很多。

**Reply by JustinJin04** [reply]
我之前没懂您的意思。后来看到wsl文档中有写[processors](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#:~:text=WSL%202%20VM.-,processors,-number)的意思是logical processor（也就是说在hyperthreading打开了以后每个物理cpu包含多个logical cpu）。所以在8个cpu的机器上需要把processors设置成16才能得到超过8x的加速

**Reply by QWenQ** [reply]
4x?文档不是要求跑到7+x吗？而且，怎么跑到7+x啊？我只跑到4x，用的i7-8750H，支持hyperthreading，lab跑在VMWare的虚拟机上，给虚拟机分了4个core

**Comment by Kriston-12** [comment]
请问一下各位佬们，第一个lab 运行 ./mandelbrot， 报错Wrote image file mandelbrot-thread.ppm，但是我看lab document里面说直接run就可以了。请问大家有遇到过这个问题吗。环境是wsl ubuntu22.04. 请各位佬帮帮孩子
$ ./mandelbrot
[mandelbrot serial]:            [311.457] ms
Wrote image file mandelbrot-serial.ppm
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
Hello world from thread 0
Hello world from thread 1
[mandelbrot thread]:            [0.140] ms
Wrote image file mandelbrot-thread.ppm
Mismatch : [0][143], Expected : 1, Actual : 0
Error : Output from threads does not match serial output

**Reply by Kriston-12** [reply]
已解决

**Comment by Nicooo-Wang** [comment]
written assignment有没有答案呀，找了全网只有楼主提供的20年3个assign有答案

**Reply by Nicooo-Wang** [reply]
找到了20年的全部作业带答案，比飞猪的全：https://www.cs.cmu.edu/afs/cs/academic/class/15418-s20/www/exercises/

**Comment by darknesswind** [comment]
15418后面要上计算集群的作业大家怎么解决的

**Comment by RisingUppercut** [comment]
同学学ucb的并行计算cs267吗，我有这个的25年视频


---- 回复的原邮件 ----
发件人  Min ***@***.***> 日期  2026年01月14日 08:37 收件人  ***@***.***> 抄送至  ***@***.***> 主题  Re: [PKUFlyingPig/cs-self-learning] CMU 15-418/Stanford CS149: Parallel Computing - CS自学指南 (Discussion #138)  

求2023的视频

**Reply by ie925155** [reply]
我想要25年的视频資源, 方便分享嗎？

**Reply by shulva** [reply]
同求

**Reply by RisingUppercut** [reply]
传到b站了

**Reply by shulva** [reply]
非常感谢(*^▽^*)

**Comment by ytoso** [comment]
高性能计算HPC

**Comment by RisingUppercut** [comment]
注意如果使用笔记本的话，做实验1要插上电源，否则speedup可能到不了文档中所说的那么高。

**Comment by EnzoDing-rgb** [comment]
这两门最新的学期都好遥远，还适合学嘛

**Comment by JFKTRSDAKIDHIKA** [comment]
现在就借助LLM去学吧





                    ***@***.***&gt;&nbsp;在 2026年4月24日 周五 10:34 写道：



 
这两门最新的学期都好遥远，还适合学嘛
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by virtualguard101** [comment]
在学CS106L的时候在课程幻灯片上看到斯坦福还有一个和并行计算相关的课程：[Stanford CME 213/ME 339](https://ericdarve.github.io/cme213-spring-2021/)，不知道质量如何，后续尝试推进一下。

**Comment by 9Skie** [comment]
很烧脑但是很有意思的课程！
